### PR TITLE
Soop TEA/PEA bugfix

### PR DESCRIPTION
**Pull request by BecCowley**
The PEA and TEA flags adjust the LATITUDE_quality_control or LONGITUDE_quality_control or TIME_quality control to value 5.
If a PER or TER is present, these flags become 3 as required.

I tried to adjust the unit test code to check, but it's failing. I know it works - just a problem with the unit tests that I can't resolve right now.

### CODE REVIEW

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 1]
```diff
@@ -469,33 +485,33 @@ def adjust_position_qc_flags(annex, data):
...
-        LOGGER.info('Date-Time failure (TER) in original file, setting time qc flag to level 4.')
-        data['TIME_quality_control'] = 4
+    if 'TE' in annex['act_code'] and 'DATI' in annex['act_parm'] and not '3' in data['TIME_quality_control']:
+        LOGGER.info('Date-Time failure (TER) in original file, setting time qc flag to level 3.')
+        data['TIME_quality_control'] = 3
         
-    if 'TE' in annex['act_code'] and ('TIME' in annex['act_parm'] or 'DATE' in annex['act_parm']):
-        LOGGER.info('Date and/or Time has been corrected (TEA) in original file, setting time qc flag to level 2.')
-        data['TIME_quality_control'] = 2
+    if 'TE' in annex['act_code'] and ('TIME' in annex['act_parm'] or 'DATE' in annex['act_parm']) and not '5' in data['TIME_quality_control']:
```
Comment (lbesnard) [inline SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:512]:
```python
    if 'TE' in annex['act_code'] and ('TIME' in annex['act_parm'] or 'DATE' in annex['act_parm']) and not data['TIME_quality_control'] == 5:
```

File: SOOP/SOOP_XBT/DELAYED/test_xbt_dm_imos_conversion.py [hunk 2]
```diff
@@ -61,18 +61,19 @@ def test_parse_annex_nc(self):
...
+        self.assertEqual(['1.0', '1.0', '1.0', '1.0', '1.0', '1.0', '1.0', '1.0', '1.0'], annex['version_soft'])
         self.assertEqual([datetime(2018, 3, 22, 0, 0), datetime(2018, 3, 22, 0, 0),
                           datetime(2018, 3, 22, 0, 0), datetime(2018, 3, 22, 0, 0),
                           datetime(2018, 3, 22, 0, 0), datetime(2018, 3, 22, 0, 0),
-                          datetime(2018, 3, 22, 0, 0), datetime(2018, 3, 22, 0, 0)], annex['prc_date'])
+                          datetime(2018, 3, 22, 0, 0), datetime(2018, 3, 22, 0, 0),
+                          datetime(2018, 3, 22, 0, 0)], annex['prc_date'])
 
-        np.testing.assert_array_almost_equal([0.67, 0.67, 1.34, 2.01, 2.68, 3.34, 274.45, 274.45], annex['aux_id'],
+        np.testing.assert_array_almost_equal([0.67, 0.67, 1.34, 2.01, 2.68, 3.34, 274.45, 274.45,0.67], annex['aux_id'],
```
Comment (lbesnard) [inline SOOP/SOOP_XBT/DELAYED/test_xbt_dm_imos_conversion.py:76]: space after coma. not an issue, but just to match the rest of the code

**Review by lbesnard — CHANGES_REQUESTED** [review]

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 3]
```diff
@@ -273,6 +282,8 @@ def parse_annex_nc(netcdf_file_path):
         prc_date = [x.replace('\x00', '') for x in prc_date]
         prc_date = list(filter(None, prc_date))
 
+        prc_date = [date.replace(' ','0') for date in prc_date]
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:285]: There are some older files with datestrings that have spaces where there should be zeros and this causes a failure when the date is read. If there is a better way to fix this, I'm happy to know what it is.

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 4]
```diff
@@ -340,13 +351,18 @@ def parse_data_nc(netcdf_file_path):
 
         no_prof, prof_type, temp_prof = temp_prof_info(netcdf_file_path)
 
-        # position QC - check this. Probably put a '1' in if the value is not set and the data has been QCd
-        if q_pos == '1':
+        # position and time QC - check this is not empty. Assume 1 if it is
+        if not q_pos:
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:355]: OK, updated. Q_pos should only ever be a single value, but changes made.

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 5]
```diff
@@ -340,13 +351,18 @@ def parse_data_nc(netcdf_file_path):
...
+            LOGGER.info('Missing position QC, flagging position with flag 1 %s' % netcdf_file_path)
             q_pos = 1
-        else:
-            q_pos = 1  # We should have flags of '1' on the lat/long, as these have been QC'd. Although not explicit in the original netcdf files (Bec Cowley 03/2020)
+        if not q_date_time:
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:358]: done

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 6]
```diff
@@ -469,33 +485,33 @@ def adjust_position_qc_flags(annex, data):
     #AW change distinguish between PE+LALO - flag =4 (position fail) and PE+LATI|LONG - flag 2 (position corrected)
     #AW we also should also set the time QC flag to 4 for date-time failures see func adjust_time_qc_flags() below
     #print("Annex=",annex)
-    if 'PE' in annex['act_code']:
+    if 'PE' in annex['act_code'] and not '5' in data['LONGITUDE_quality_control']:
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:488]: Updated and tested in unittest.

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 7]
```diff
@@ -469,33 +485,33 @@ def adjust_position_qc_flags(annex, data):
...
             #print("annex['act_parm']1=",annex['act_parm'])
-            LOGGER.info('Position correction (PEA) in original file, changing position flags to level 2.')
+            LOGGER.info('Position correction (PEA) in original file, changing position flags to level 5.')
 
-            data['LATITUDE_quality_control'] = 2
-            data['LONGITUDE_quality_control'] = 2
-        if 'LALO' in annex['act_parm']:
+            data['LATITUDE_quality_control'] = 5
+            data['LONGITUDE_quality_control'] = 5
+        if 'LALO' in annex['act_parm'] and not '3' in data['LONGITUDE_quality_control']:
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:496]: Updated and added test in test_xbt_dm_imos_conversion.py

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py [hunk 8]
```diff
@@ -469,33 +485,33 @@ def adjust_position_qc_flags(annex, data):
...
 
 def adjust_time_qc_flags(annex, data):
     #AW Add function  we also should also set the time QC flag to 4 for date-time failures TE in annex['act_code'] + DATI in annex['act_parm']
-    #or set time QC to flag 2 if date/time has been corrected
+    #or set time QC to flag 5 if date/time has been corrected
     #print("Annex=",annex)
-    if 'TE' in annex['act_code'] and 'DATI' in annex['act_parm']:
-        LOGGER.info('Date-Time failure (TER) in original file, setting time qc flag to level 4.')
-        data['TIME_quality_control'] = 4
+    if 'TE' in annex['act_code'] and 'DATI' in annex['act_parm'] and not '3' in data['TIME_quality_control']:
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:508]: done

[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:508]: Will need to find a file with a TE flag to test this bit of code on.

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py (hunk 1 above)
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:512]: done, will need to test on a file with TE flag.

File: SOOP/SOOP_XBT/DELAYED/test_xbt_dm_imos_conversion.py [hunk 9]
```diff
@@ -125,19 +126,20 @@ def test_parse_edited_nc_netcdf_test_1(self):
...
-        self.assertEqual(['CS', 'CS', 'CS', 'CS', 'CS', 'CS', 'CS', 'CS'], annex['ident_code'])
+        self.assertEqual(['TEMP', 'TEMP', 'TEMP', 'TEMP', 'TEMP', 'TEMP', 'TEMP', 'TEMP', 'LATI'], annex['act_parm'])
+        self.assertEqual(['QC', 'CS', 'CS', 'CS', 'CS', 'CS', 'HB', 'NG', 'PE'], annex['act_code'])
+        self.assertEqual(['CSCB', 'CSCB', 'CSCB', 'CSCB', 'CSCB', 'CSCB', 'CSCB', 'CSCB', 'CSCB'], annex['prc_code'])
+        self.assertEqual(['CS', 'CS', 'CS', 'CS', 'CS', 'CS', 'CS', 'CS', 'CS'], annex['ident_code'])
 
-        self.assertEqual([25.099, 25.099, 25.112, 25.12, 25.125, 25.128, 12.253, 12.253], annex['previous_val'])
-        self.assertEqual(['1.0', '1.0', '1.0', '1.0', '1.0', '1.0', '1.0', '1.0'], annex['version_soft'])
+        self.assertEqual([25.099, 25.099, 25.112, 25.12, 25.125, 25.128, 12.253, 12.253, -34.1], annex['previous_val'])
+        self.assertEqual(['1.0', '2.0', '2.0', '2.0', '2.0', '2.0', '1.0', '1.0', '1.0'], annex['version_soft'])
```
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/test_xbt_dm_imos_conversion.py:135]: Done and fixed.

**Review by BecCowley — COMMENTED** [review]
@lbesnard I've made some of the changes suggested and commented on others.
I've made the changes in the SOOP_TEA_PEAbugfix branch. Do I re-do a pull request with the updates?

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lbesnard) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:285]: >  I've made the changes in the SOOP_TEA_PEAbugfix branch. Do I re-do a pull request with the updates?

@BecCowley Thanks for this. You just have to push the commit with your changes online and it should be automatically picked up

[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:284]: @lbesnard I have pushed the changes to my fork - is that all that is required?

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:353]: I have reverted as q_pos is read in as a byte. I now change it to an integer. See latest commit.

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py:356]: I have reverted as q_date_time is read in as a byte. I change it to an integer. See latest commit.

File: SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py (hunk 1 above)
Comment (BecCowley) [reply SOOP/SOOP_XBT/DELAYED/xbt_dm_imos_conversion.py]: Missed one of these, recommitted.

**Review by lbesnard — APPROVED** [review]
@BecCowley Thanks for the changes. It looks good. I'm glad you're getting used to the unittests :)
I"ll merge once you confirm it's all good to go in case you're still adding some commits in the same branch

**Comment by BecCowley** [comment]
@lbesnard I think we can go ahead and merge. I'm happy we caught the PEA/TEA changes correctly. There will be more updates as I dig backwards in time to older and trickier data.

**Pull request merged**